### PR TITLE
trie implementation

### PR DESCRIPTION
**Pull request by sdonkov**
Trie implementation

### CODE REVIEW

File: src/Demo.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
+import java.util.ArrayList;
+
+public class Demo {
```
Comment (dmitankin) [inline src/Demo.java:3]: It's a good practice to put TODO comments when you are checking in something you deem temporary.
In this case it'd be something like `// TODO replace with a proper test`

File: src/PrefixTree.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        TraverseResult traverseResult = new TraverseResult();
+        traverseResult.node = root;
+        for (; traverseResult.index < word.length(); traverseResult.index++) {
+            ArrayList<Edge> nodeEdges = traverseResult.node.getEdges();
+
+            Optional<Edge> edgeFound = nodeEdges.stream()
+                    .filter(edge -> edge.getNode().getValue() == word.charAt(traverseResult.index))
+                    .findFirst();
+
+            if (edgeFound.isEmpty()) {
```
Comment (dmitankin) [inline src/PrefixTree.java]: There is no `isEmpty` in Java 8 for `Optional`. It's introduced in Java 11.

File: src/Node.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+    Node(char value) {
+        this.value = value;
+    }
+
+    public boolean isFinal() {
+        return isFinal;
+    }
+
+    public void setFinal(boolean aFinal) {
```
Comment (dmitankin) [inline src/Node.java]: What is this `a` in front of the param for? It implies some notation that is not common.
You can label it `value` instead, but you already have such a property so it'd be confusing too.
I guess the proper solution is to avoid keywords because you are overloading the terms this way. This may become end-of-word instead.

File: src/Node.java [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+        isFinal = aFinal;
+    }
+
+    public char getValue() {
+        return value;
+    }
+
+    public ArrayList<Edge> getEdges() {
```
Comment (dmitankin) [inline src/Node.java]: If the consumer does not depend on this being an ArrayList (and I think they don't), this should be simply a `List`. Whether it's an ArrayList is an implementation detail. Interfaces are the way to go when you are describing a public API. They allow for decoupling and give the library maintainers the freedom to change implementation when needed without compromising the contract of the API, i.e. without introducing breaking changes.
You might even go further and narrow this down to `Collection`. It depends on whether consumers depend on what `List` brings to the table (which is order).
The golden rule is to expose as little as possible for the consumers to do their job.

File: src/Node.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+        return edges;
+    }
+
+    public void addEdge(Edge edge) {
+        edges.add(edge);
+    }
```
Comment (dmitankin) [inline src/Node.java]: This is a bit controversial as it implies you can't add to the list being returned above (which is not true). I guess the proper way would be to return an unmodifiable collection from `getEdges` and expose adding with this method.
Implementing your own `add` makes sense in order to prevent adding the same letter edge more than once. This may be left for a future change along with the tests which will expose all kinds of edge cases (pun not intended).

File: src/Edge.java [hunk 6]
```diff
@@ -0,0 +1,19 @@
+public class Edge {
```
Comment (dmitankin) [inline src/Edge.java]: This whole class now looks redundant for the trie. We'll need it for the dawg, I guess, but you can nuke it from here.

File: src/PrefixTree.java [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+    private static class TraverseResult {
+        private Node node;
+        private int index;
+
+    }
+
+    public boolean contains(CharSequence word) {
+        TraverseResult traverseResult = traverse(word);
+        if (traverseResult.index == word.length() && traverseResult.node.isFinal()) {
+            System.out.println("Tree contains this word - " + word);
```
Comment (dmitankin) [inline src/PrefixTree.java:63]: The `println` calls are all debug lines which won't stay so thay too should get a TODO (see #4) or simply be removed.

**Review by dmitankin — COMMENTED** [review]
No need to use underscores in the branch name. They look machine-like and are harder to type than dashes.

**Review by dmitankin — COMMENTED** [review]
Please remove the class files from the PR (perhaps add them to your .gitignore too).
It's a standard practice to self-review your change before publishing it for review in order to avoid such mishaps.

File: .idea/misc.xml [hunk 8]
```diff
@@ -7,7 +7,7 @@
       </list>
     </option>
   </component>
-  <component name="ProjectRootManager" version="2" languageLevel="JDK_18" default="true" project-jdk-name="18" project-jdk-type="JavaSDK">
+  <component name="ProjectRootManager" version="2" languageLevel="JDK_1_8" project-jdk-name="18" project-jdk-type="JavaSDK">
```
Comment (dmitankin) [inline .idea/misc.xml]: This whole file should go the way of the dodo too.

File: src/Node.java [hunk 9]
```diff
@@ -0,0 +1,42 @@
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+
+public class Node {
```
Comment (dmitankin) [inline src/Node.java]: There's an oportunity to reduce the contract area even further. This class should not be public really. It's not part of the API of the class after all, i.e. it's really an something internal to the trie - it exists solely to simplify the implementation.
It's small and simple enough to even move it within the PrefixTree file.

File: src/PrefixTree.java [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+    private final Node root;
+
+    PrefixTree() {
+        this.root = new Node('\u0000');
+    }
+
+    private static class TraverseResult {
+        private Node node;
+        private int index;
+
```
Comment (dmitankin) [inline src/PrefixTree.java:49]: Redundant line.

File: src/Node.java [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+        this.endOfWord = endOfWord;
+    }
+
+    public char getValue() {
+        return value;
+    }
+
+    public Collection<Node> getEdges() {
+        return Collections.unmodifiableCollection(edges);
+    }
```
Comment (dmitankin) [inline src/Node.java]: Thinking of reducing this even further, we don't need a collection here really. All we care about is finding the edge for a particular character. It makes sense to move the search here (the lambda) and transform this into a method that returns an `Optional<Node>` for a given character.

**Review by dmitankin — COMMENTED** [review]
I think this'd be the final review, but let's see...

File: src/PrefixTree.java [hunk 12]
```diff
@@ -13,7 +14,47 @@ public class PrefixTree {
...
+        public void setEndOfWord(boolean endOfWord) {
+            this.endOfWord = endOfWord;
+        }
+
+        public char getValue() {
+            return value;
+        }
+
+        public Optional<Node> findChild(char searchedChar) {
```
Comment (dmitankin) [inline src/PrefixTree.java]: In the semantics of the getter above, this parameter should be called `value`.

File: src/PrefixTree.java [hunk 13]
```diff
@@ -13,7 +14,47 @@ public class PrefixTree {
     private static class TraverseResult {
         private Node node;
         private int index;
+    }
+
+    private static class Node {
+
+        private boolean endOfWord;
+        private final char value;
+        private ArrayList<Node> edges = new ArrayList<>();
```
Comment (dmitankin) [inline src/PrefixTree.java]: This looks like a `final` field.

File: src/PrefixTree.java [hunk 14]
```diff
@@ -13,7 +14,47 @@ public class PrefixTree {
...
+            this.value = value;
+        }
+
+        public boolean isEndOfWord() {
+            return endOfWord;
+        }
+
+        public void setEndOfWord(boolean endOfWord) {
+            this.endOfWord = endOfWord;
+        }
```
Comment (dmitankin) [inline src/PrefixTree.java]: Given that there's no remove operation, this can lose the parameter and become an idempotent way of marking a node as final.

File: src/PrefixTree.java [hunk 15]
```diff
@@ -13,7 +14,47 @@ public class PrefixTree {
...
+
+        public Optional<Node> findChild(char searchedChar) {
+            return this.edges.stream()
+                    .filter(edge -> edge.getValue() == searchedChar)
+                    .findFirst();
+        }
+
+        public void addEdge(Node node) {
+            edges.add(node);
+        }
```
Comment (dmitankin) [inline src/PrefixTree.java:48]: Future improvement:
I wonder about this one. You can make a node point to itself and all other kind of crazy edge cases that we'll have to deal with. Won't it be better if it accepts a character and returns a new Node that's a child of this one? So `Node createChild(char value)` like. Perhaps it can even return a node if the child already exists thus removing the need to use `traverse` for adding which will get us rid of the TraverseResult. It's a big change, so it deserves its own commit after we have proper validation and some performance tests enabled.

**Review by dmitankin — COMMENTED** [review]
Almost there...

**Review by dmitankin — APPROVED** [review]
Looks good, thanks!
